### PR TITLE
Fix/registry oauth2

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR introduces OCI Registry HTTP API V2 authorization via OAuth for 

- pulling `jujud-operator`, `juju-db` and `charm-base` images from k8s controllers, application operators, and sidecar pods;
- fetching available controller versions to upgrade for `upgrade-controller` command;

The tested registries are(the rest of registries will be implemented and tested in following PRs):

| Registry | Public | Private |
| --- | --- | --- |
| azurecr.io |  |  |
| docker.io | :white_check_mark: | :white_check_mark: |
| ecr.aws |  |  |
| gcr.io |  |  |
| ghcr.io |  | :white_check_mark: |
| registry.gitlab.com |  | :white_check_mark: |
| quay.io |  |  |

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
# private registries:

# prepare credentials like below (you can grab these credentials from `~/.docker/config.json` once `docker login` succeeded)
$ cat dockerhub.json
{
    // "auth": "xxxxx==",
    "repository": "jujuqa",
    "username": "jujuqa",
    "password": "xxx",
}

$ cat ghcr.json
{
    "auth": "xxxx",
    "repository": "ghcr.io/jujuqa",
    "serveraddress": "ghcr.io",
}

$ cat gitlab.json
{
    "repository": "registry.gitlab.com/jujuqa",
    "serveraddress": "registry.gitlab.com",
    "username": "jujuqa",
    "password": "xxxx",
}

$ juju bootstrap microk8s k1 --config caas-image-repo="'$(cat dockerhub.json)'" --debug

$ juju upgrade-controller --agent-stream=develop

# public registry

$ juju bootstrap microk8s k1 --config caas-image-repo="ghcr.io/jujuqa"

$ juju upgrade-controller --agent-stream=develop

```

## Documentation changes

Yes

## Bug reference

https://bugs.launchpad.net/juju/+bug/1935830
https://bugs.launchpad.net/juju/+bug/1940820
https://bugs.launchpad.net/juju/+bug/1935953

### CODE REVIEW

**Comment by ycliuhw** [comment]
[OCI registry API supports v2 bearer token oauth](https://trello.com/c/ynqp6VVP/1729-oci-registry-api-supports-v2-bearer-token-oauth)

File: docker/registry/acr.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/docker"
+)
+
+type acr struct {
+	*baseClient
+}
+
+func newACR(repoDetails docker.ImageRepoDetails, transport http.RoundTripper) RegistryInternal {
```
Comment (hpidcock) [inline docker/registry/acr.go:20]: rename to something easier to identitfy like newAzureACR or newAzureContainerRegistry

File: docker/registry/gcr.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/docker"
+)
+
+type gcr struct {
+	*baseClient
+}
+
+func newGCR(repoDetails docker.ImageRepoDetails, transport http.RoundTripper) RegistryInternal {
```
Comment (hpidcock) [inline docker/registry/gcr.go:19]: rename to newGoogleGCR or newGoogleContainerRegistry

File: docker/registry/gitlab.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	"strings"
+
+	"github.com/juju/juju/docker"
+)
+
+type gitlab struct {
+	*baseClient
+}
+
+func newGitlab(repoDetails docker.ImageRepoDetails, transport http.RoundTripper) RegistryInternal {
```
Comment (hpidcock) [inline docker/registry/gitlab.go:17]: newGitLabContainerRegistry

File: docker/registry/transports.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	if t.AccessToken != "" {
+		return t.AccessToken
+	}
+	if t.Token != "" {
+		return t.Token
+	}
+	return ""
+}
+
+func (t *tokenTransport) refreshoAuthToken(failedResp *http.Response) error {
```
Comment (hpidcock) [inline docker/registry/transports.go:123]: Use a capital O in oauth. It reads like Refresho Auth Token

File: docker/registry/registry.go [hunk 5]
```diff
@@ -24,123 +24,152 @@ import (
...
+	url := *c.baseURL
+	ver := c.repoDetails.APIVersion().String()
 	if !strings.HasSuffix(strings.TrimRight(url.Path, "/"), ver) {
 		url.Path = path.Join(url.Path, ver)
 	}
 	if url.Scheme == "" {
 		url.Scheme = "https"
 	}
 	url.Path = path.Join(url.Path, pathSuffix)
+	logger.Criticalf("baseClient url.Path ===> %q, %q", url.Path, pathSuffix)
```
Comment (hpidcock) [inline docker/registry/registry.go:147]: obviously these need to be removed/made tracef

**Review by hpidcock — COMMENTED** [review]
Looks promising

File: docker/registry/transports.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+// RoundTrip executes a single HTTP transaction, returning a Response for the provided Request.
+func (t *tokenTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if err := t.authorizeRequest(req); err != nil {
+		return nil, errors.Trace(err)
+	}
+	resp, err := t.transport.RoundTrip(req)
+	if err != nil || resp == nil || resp.StatusCode != http.StatusUnauthorized {
+		return resp, errors.Trace(err)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply docker/registry/transports.go:187]: right

File: docker/registry/transports.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+	transport http.RoundTripper
+	username  string
+	password  string
+	authToken string
+}
+
+func newBasicTransport(
+	transport http.RoundTripper, username string, password string, authToken string,
+) http.RoundTripper {
+	logger.Criticalf("newBasicTransport %q, %q, %q", username, password, authToken)
```
[the remark this answers is not in the document]
Comment (ycliuhw) [reply docker/registry/transports.go:29]: Yeah, currently still doing some testing. Will remove it later.

File: controller/config.go [hunk 8]
```diff
@@ -857,11 +857,12 @@ func validateCAASImageRepo(imageRepo string) (string, error) {
 		return "", errors.Trace(err)
 	}
 	if imageDetails.IsPrivate() {
-		r, err := registry.NewRegistry(*imageDetails)
+		r, err := registry.New(*imageDetails)
 		if err != nil {
 			return "", errors.Trace(err)
 		}
 		*imageDetails = r.ImageRepoDetails()
+		func() { _ = r.Close() }()
```
Comment (wallyworld) [inline controller/config.go:865]: why the func?

File: docker/auth.go [hunk 9]
```diff
@@ -166,13 +155,8 @@ func (rid *ImageRepoDetails) init() error {
...
 	return nil
 }
-
-func (rid ImageRepoDetails) APIVersion() APIVersion {
-	v := APIVersionV1
-	if !rid.TokenAuthConfig.Empty() {
-		v = APIVersionV2
-	}
-	return v
+func (rid ImageRepoDetails) Empty() bool {
```
Comment (wallyworld) [inline docker/auth.go:158]: should comment

File: docker/registry/acr.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+type azureContainerRegistry struct {
+	*baseClient
+}
+
+func newAzureContainerRegistry(repoDetails docker.ImageRepoDetails, transport http.RoundTripper) RegistryInternal {
+	c := newBase(repoDetails, DefaultTransport)
+	return &azureContainerRegistry{c}
+}
+
+func (c *azureContainerRegistry) Match() bool {
```
Comment (wallyworld) [inline docker/registry/acr.go:27]: should comment these

File: docker/registry/acr.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+	if len(parts) < 1 {
+		return "", errors.NotValidf("auth %q", auth)
+	}
+	return parts[0], nil
+}
+
+func (c *azureContainerRegistry) WrapTransport() error {
+	transport := c.client.Transport
+	if c.repoDetails.IsPrivate() {
+		if !c.repoDetails.TokenAuthConfig.Empty() {
```
Comment (wallyworld) [inline docker/registry/acr.go:47]: combine into one if {} ?

File: docker/registry/base.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+					c.repoDetails.ServerAddress,
+				),
+			)
+		}
+	}
+	c.client.Transport = newErrorTransport(transport)
+	return nil
+}
+
+func (c *baseClient) DecideBaseURL() error {
```
Comment (wallyworld) [inline docker/registry/base.go:112]: comment public functions

File: docker/registry/dockerhub.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+	c := newBase(repoDetails, DefaultTransport)
+	return &dockerhub{c}
+}
+
+func (c *dockerhub) Match() bool {
+	c.prepare()
+	return c.repoDetails.ServerAddress == "" || strings.Contains(c.repoDetails.ServerAddress, "docker.io")
+}
+
+func (c *dockerhub) DecideBaseURL() error {
```
Comment (wallyworld) [inline docker/registry/dockerhub.go:34]: `DecideBaseURL` seems like it should be called something like `Initialise` or something. 
Actually, the functionality here seems to pertain to the `repoDetails` value - should this be an internal method used to set up things whenever the `repoDetails` is set, ie should this be called from `newDockerHub` and not exposed as a separate method?

**Review by wallyworld — APPROVED** [review]
One change I think should be that `DecideBaseURL` is an internal detail of initialising each repo, not exposed as a separate function.

File: docker/registry/dockerhub.go (hunk 13 above)
Comment (ycliuhw) [reply docker/registry/dockerhub.go:34]: DecideBaseURL is exported for generating mocks and it's one of the methods called in `initClient`.
So DecideBaseURL is just one of the steps of Initialise.
I will move the provider integration code to `internal` sub-package in the following PR.

File: controller/config.go (hunk 8 above)
Comment (ycliuhw) [reply controller/config.go:865]: Ah, I think I missed defer, but it doesn't really matter.

**Comment by ycliuhw** [comment]
!!build!!

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[integrate more registry support, for example gcr, azure, etc.](https://trello.com/c/K1isSDRz/1749-integrate-more-registry-support-for-example-gcr-azure-etc)

**Pull request merged**